File: frontend/src/lib/types/permissions.ts
/**
 * Mirrors backend/app/core/permissions.py exactly — the capability matrix here
 * must stay in sync with ROLE_PERMISSIONS server-side. This only gates what the
 * UI shows; the API re-checks every one of these server-side regardless, so a
 * drift here is a UX bug (wrong nav item visible/hidden), never a security hole.
 */

export type PoliceRank =
  | "CONSTABLE"
  | "INSPECTOR"
  | "DSP"
  | "SP"
  | "DGP"
  | "CRIME_ANALYST"
  | "POLICY_MAKER"

export type Permission =
  | "view_case_basic"
  | "view_case_sensitive"
  | "edit_case"
  | "verify_person"
  | "view_pii_unmasked"
  | "upload_document"
  | "promote_document"
  | "view_network_basic"
  | "view_network_advanced"
  | "view_trends_hotspots"
  | "compute_risk_score"
  | "review_risk_score"
  | "view_financial_raw"
  | "generate_case_brief"
  | "view_aggregate_analytics"
  | "export_report"
  | "share_case"
  | "view_audit_log"
  | "manage_users"
  | "manage_case_notes"
  | "manage_analytics_jobs"

const CONSTABLE_TIER: Permission[] = ["view_case_basic", "view_trends_hotspots"]

const INSPECTOR_TIER: Permission[] = [
  ...CONSTABLE_TIER,
  "view_case_sensitive",
  "view_pii_unmasked",
  "view_network_basic",
  "generate_case_brief",
]

const DSP_TIER: Permission[] = [
  ...INSPECTOR_TIER,
  "edit_case",
  "verify_person",
  "upload_document",
  "promote_document",
  "view_network_advanced",
  "compute_risk_score",
  "review_risk_score",
  "view_financial_raw",
  "view_aggregate_analytics",
  "export_report",
  "share_case",
  "manage_case_notes",
]

const SP_TIER: Permission[] = [...DSP_TIER, "view_audit_log"]
const DGP_TIER: Permission[] = [...SP_TIER, "manage_users", "manage_analytics_jobs"]

const CRIME_ANALYST_TIER: Permission[] = [
  ...DSP_TIER.filter((permission) => permission !== "share_case" && permission !== "manage_case_notes"),
  "manage_analytics_jobs",
]

const POLICY_MAKER_TIER: Permission[] = [
  "view_case_basic",
  "view_trends_hotspots",
  "view_aggregate_analytics",
  "view_audit_log",
]

export const ROLE_PERMISSIONS: Record<PoliceRank, Permission[]> = {
  CONSTABLE: CONSTABLE_TIER,
  INSPECTOR: INSPECTOR_TIER,
  DSP: DSP_TIER,
  SP: SP_TIER,
  DGP: DGP_TIER,
  CRIME_ANALYST: CRIME_ANALYST_TIER,
  POLICY_MAKER: POLICY_MAKER_TIER,
}

export function roleHasPermission(role: PoliceRank | undefined, permission: Permission): boolean {
  if (!role) return false
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false
}

export function roleHasAnyPermission(role: PoliceRank | undefined, permissions: Permission[]): boolean {
  return permissions.some((permission) => roleHasPermission(role, permission))
}

export const RANK_LABELS: Record<PoliceRank, string> = {
  CONSTABLE: "Constable",
  INSPECTOR: "Inspector",
  DSP: "DSP",
  SP: "SP",
  DGP: "DGP",
  CRIME_ANALYST: "Crime Analyst",
  POLICY_MAKER: "Policy Maker",
}

/**
 * Plain-language copy for the public access-scope page. English-only on
 * purpose, matching RANK_LABELS above: these name fixed capability and rank
 * identifiers rather than prose, and the surrounding page chrome is what goes
 * through i18n.
 */
export const RANK_SUMMARIES: Record<PoliceRank, string> = {
  CONSTABLE: "Read-only access to non-sensitive case records and public crime trends.",
  INSPECTOR: "Adds sensitive case detail, unmasked identities, and basic network views.",
  DSP: "Full investigative access, including editing, risk scoring, and financial records.",
  SP: "Everything a DSP can do, plus visibility of the governance audit log.",
  DGP: "Complete access, and the only rank that can administer user accounts.",
  CRIME_ANALYST: "Investigative and analytics access without supervisory sign-off actions.",
  POLICY_MAKER: "Aggregate statistics only, with no access to individual case or person data.",
}

export const PERMISSION_LABELS: Record<Permission, string> = {
  view_case_basic: "View basic case records",
  view_case_sensitive: "View sensitive case detail",
  edit_case: "Create and edit cases",
  manage_case_notes: "Moderate other officers' case notes",
  share_case: "Share a case with another unit",
  generate_case_brief: "Generate AI case briefs",
  verify_person: "Verify person identities",
  view_pii_unmasked: "View unmasked personal data",
  upload_document: "Upload case documents",
  promote_document: "Promote documents to evidence",
  view_network_basic: "View basic criminal networks",
  view_network_advanced: "View advanced network analysis",
  compute_risk_score: "Run recidivism risk scoring",
  review_risk_score: "Review and sign off risk scores",
  view_financial_raw: "View raw financial records",
  view_trends_hotspots: "View crime trends and hotspots",
  view_aggregate_analytics: "View district aggregate analytics",
  export_report: "Export and download reports",
  view_audit_log: "View the governance audit log",
  manage_analytics_jobs: "Trigger analytics and model jobs",
  manage_users: "Manage user accounts and roles",
}

/** Ordering and grouping for the access-scope matrix, coarse to sensitive. */
export const PERMISSION_GROUPS: { title: string; permissions: Permission[] }[] = [
  {
    title: "Cases and records",
    permissions: [
      "view_case_basic",
      "view_case_sensitive",
      "edit_case",
      "generate_case_brief",
      "share_case",
      "manage_case_notes",
    ],
  },
  { title: "People and identity", permissions: ["verify_person", "view_pii_unmasked"] },
  { title: "Documents", permissions: ["upload_document", "promote_document"] },
  { title: "Network analysis", permissions: ["view_network_basic", "view_network_advanced"] },
  {
    title: "Risk and financial",
    permissions: ["compute_risk_score", "review_risk_score", "view_financial_raw"],
  },
  {
    title: "Analytics and reporting",
    permissions: ["view_trends_hotspots", "view_aggregate_analytics", "export_report"],
  },
  {
    title: "Governance",
    permissions: ["view_audit_log", "manage_analytics_jobs", "manage_users"],
  },
]

export const ALL_RANKS: PoliceRank[] = [
  "CONSTABLE",
  "INSPECTOR",
  "DSP",
  "SP",
  "DGP",
  "CRIME_ANALYST",
  "POLICY_MAKER",
]

File: backend/app/core/permissions.py
"""
Capability-based RBAC (§10 of the design doc, "Secure Role-Based Access & Governance").

The seven ranks in app.models.enums.Role don't form one clean chain — CRIME_ANALYST
and POLICY_MAKER are specialist tracks, not command rungs — so access is expressed as
named capabilities per role rather than a single numeric hierarchy check. Every grant
below traces back to a concrete existing endpoint (see the historical require_roles(...)
call sites this replaces), plus the handful of new capabilities the governance features
in this pass introduce (edit_case, export_report, view_audit_log, manage_users, ...).
"""

import enum
from typing import FrozenSet

from fastapi import Depends, HTTPException, status

from app.core.security import get_current_user
from app.models.enums import Role
from app.models.user import User


class Permission(str, enum.Enum):
    VIEW_CASE_BASIC = "view_case_basic"
    VIEW_CASE_SENSITIVE = "view_case_sensitive"          # brief_facts, unmasked identity fields
    EDIT_CASE = "edit_case"
    VERIFY_PERSON = "verify_person"
    VIEW_PII_UNMASKED = "view_pii_unmasked"
    UPLOAD_DOCUMENT = "upload_document"
    PROMOTE_DOCUMENT = "promote_document"
    VIEW_NETWORK_BASIC = "view_network_basic"            # ego network, co-offending, predicted links
    VIEW_NETWORK_ADVANCED = "view_network_advanced"      # community detection, multi-jurisdiction
    VIEW_TRENDS_HOTSPOTS = "view_trends_hotspots"
    COMPUTE_RISK_SCORE = "compute_risk_score"
    REVIEW_RISK_SCORE = "review_risk_score"
    VIEW_FINANCIAL_RAW = "view_financial_raw"
    GENERATE_CASE_BRIEF = "generate_case_brief"
    VIEW_AGGREGATE_ANALYTICS = "view_aggregate_analytics"  # district-year socio/crime-stat aggregates only
    EXPORT_REPORT = "export_report"
    SHARE_CASE = "share_case"
    VIEW_AUDIT_LOG = "view_audit_log"
    MANAGE_USERS = "manage_users"
    MANAGE_CASE_NOTES = "manage_case_notes"              # delete/moderate notes authored by someone else
    MANAGE_ANALYTICS_JOBS = "manage_analytics_jobs"      # trigger GWR recompute / embedding backfill batch jobs


_CONSTABLE_TIER: FrozenSet[Permission] = frozenset(
    {Permission.VIEW_CASE_BASIC, Permission.VIEW_TRENDS_HOTSPOTS}
)

_INSPECTOR_TIER: FrozenSet[Permission] = _CONSTABLE_TIER | {
    Permission.VIEW_CASE_SENSITIVE,
    Permission.VIEW_PII_UNMASKED,
    Permission.VIEW_NETWORK_BASIC,
    Permission.GENERATE_CASE_BRIEF,
}

_DSP_TIER: FrozenSet[Permission] = _INSPECTOR_TIER | {
    Permission.EDIT_CASE,
    Permission.VERIFY_PERSON,
    Permission.UPLOAD_DOCUMENT,
    Permission.PROMOTE_DOCUMENT,
    Permission.VIEW_NETWORK_ADVANCED,
    Permission.COMPUTE_RISK_SCORE,
    Permission.REVIEW_RISK_SCORE,
    Permission.VIEW_FINANCIAL_RAW,
    Permission.VIEW_AGGREGATE_ANALYTICS,
    Permission.EXPORT_REPORT,
    Permission.SHARE_CASE,
    Permission.MANAGE_CASE_NOTES,
}

_SP_TIER: FrozenSet[Permission] = _DSP_TIER | {Permission.VIEW_AUDIT_LOG}

_DGP_TIER: FrozenSet[Permission] = _SP_TIER | {Permission.MANAGE_USERS, Permission.MANAGE_ANALYTICS_JOBS}

# CRIME_ANALYST mirrors the data-steward capabilities of DSP-tier (they did this work
# under the old ANALYST role) without the command-only actions (SHARE_CASE — a
# supervisory sign-off — and MANAGE_CASE_NOTES moderation).
_CRIME_ANALYST_TIER: FrozenSet[Permission] = (
    _DSP_TIER - {Permission.SHARE_CASE, Permission.MANAGE_CASE_NOTES}
) | {Permission.MANAGE_ANALYTICS_JOBS}

# POLICY_MAKER is deliberately aggregate-only — no case PII, no network graph, no
# case editing. This is the design doc §6.1 ecological-fallacy/individual-data wall,
# expressed as a permission grant rather than a data-layer check.
_POLICY_MAKER_TIER: FrozenSet[Permission] = frozenset(
    {
        Permission.VIEW_CASE_BASIC,
        Permission.VIEW_TRENDS_HOTSPOTS,
        Permission.VIEW_AGGREGATE_ANALYTICS,
        Permission.VIEW_AUDIT_LOG,
    }
)

ROLE_PERMISSIONS: dict[Role, FrozenSet[Permission]] = {
    Role.CONSTABLE: _CONSTABLE_TIER,
    Role.INSPECTOR: _INSPECTOR_TIER,
    Role.DSP: _DSP_TIER,
    Role.SP: _SP_TIER,
    Role.DGP: _DGP_TIER,
    Role.CRIME_ANALYST: _CRIME_ANALYST_TIER,
    Role.POLICY_MAKER: _POLICY_MAKER_TIER,
}


def role_has_permission(role: Role, permission: Permission) -> bool:
    return permission in ROLE_PERMISSIONS.get(role, frozenset())


def require_permission(*permissions: Permission):
    """
    FastAPI dependency factory — grants access if the caller's role holds ANY of the
    listed permissions (mirrors the historical require_roles(*roles) semantics).
    """

    async def _check(current_user: User = Depends(get_current_user)) -> User:
        granted = ROLE_PERMISSIONS.get(current_user.role, frozenset())
        if not granted.intersection(permissions):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Requires one of permissions: {[p.value for p in permissions]}",
            )
        return current_user

    return _check
